components/hab/src/command/pkg/list.rs

```diff
@@ -0,0 +1,94 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common::ui::{Status, UIWriter, UI};
+use error::{Error, Result};
+use hcore::fs as hfs;
+use hcore::package::ident;
+use hcore::package::list;
+use hcore::package::PackageIdent;
+use std::cmp::Ordering;
+use std::fmt;
+use std::path::Path;
+use std::str::FromStr;
+
+/// There are three options for what we can list:
+///   - All packages (no prefix supplied)
+///   - All packages in an origin (string with no '/')
+///   - A set of packages with a common package ident (e.g. /ORIGIN/NAME) with optionally version/release
+#[derive(Debug)]
+enum Listing {
+    AllPackages,
+    Origin(String),
+    Ident(PackageIdent),
+}
+
+impl fmt::Display for Listing {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Listing::AllPackages => write!(f, "all packages"),
+            Listing::Origin(ref o) => write!(f, "packages for origin {}", o),
+            Listing::Ident(ref p) => write!(f, "packages for {}", p),
+        }
+    }
+}
+
+pub fn start(ui: &mut UI, prefix: &Option<String>, fs_root_path: &Path) -> Result<()> {
+    let listing = listing_type_for_prefix(prefix)?;
+    ui.status(Status::Determining, format!("{}", &listing))?;
+    let package_path = hfs::pkg_root_path(Some(&fs_root_path));
+
+    let mut packages = match listing {
+        Listing::AllPackages => list::all_packages(&package_path)?,
+        Listing::Origin(origin) => list::package_list_for_origin(&package_path, &origin)?,
+        Listing::Ident(ident) => list::package_list_for_ident(&package_path, &ident)?,
+    };
+
+    packages.sort_unstable_by(|a, b| component_cmp(a, b));
+    for p in &packages {
+        println!("{}", &p);
+    }
+
+    Ok(())
+}
+
+fn listing_type_for_prefix(prefix: &Option<String>) -> Result<Listing> {
+    match prefix {
+        None => Ok(Listing::AllPackages),
+        Some(p) => {
+            if ident::is_valid_origin_name(&p) {
+                Ok(Listing::Origin(p.clone()))
+            } else {
+                match PackageIdent::from_str(&p) {
+                    Ok(ident) => Ok(Listing::Ident(ident)),
+                    Err(e) => Err(Error::HabitatCore(e)),
+                }
+            }
+        }
+    }
+}
+
+/// Compare two `PackageIdent`s component by component:
+/// i.e. start with origin, then name, then version, then
+/// release. The first component to be not equal, then return
+/// the greater/lesser
+fn component_cmp(a: &PackageIdent, b: &PackageIdent) -> Ordering {
```

Comment:
Why not add this to the core implementation of `PackageIdent`? In fact, it might be possible to just add `#[derive(Ord, PartialOrd)]` to get the desired behavior with even less code. Then you can replace the `packages.sort_unstable_by(…)` above with just `packages.sort()`.

Reply:
`PackageIdent` already has `Ord` and `PartialOrd` but it's a different implementation which only compares the last 2 segments of a package ident - https://github.com/habitat-sh/core/blob/master/components/core/src/package/ident.rs#L226-L311

Happy to move it over if the fn is more generally useful but I guess I'd need a better name and it'd still be called via `sort_unstable_by` ?

Reply:
Huh, that seems like a surprising default for `Ord` that comparing `my-origin/redis` and `core/redis` are equal. I'm guessing there's a need for that, but to my mind *that*'s the situation that needs a special comparison function and merely sorting a list of packages for display should use a default which considers all the fields. Maybe @fnichol can provide more insight, but the origin-ignoring comparisons seem like a pretty un-intuitive and surprising default to me.

Reply:
Yeah.  I modified the `Ord::cmp` for `PackageIdent` and no tests failed in Habitat, so I'm not touching the refactor of `Ord`/`PartialOrd` until I understand what it's supposed to be doing !

:+1: on moving the function into `PackageIdent` as first steps

Reply:
@baumanj Updated the code and it looks a lot cleaner.  Requires habitat-sh/core#84. Will update Cargo.lock when that is merged